fawltydeps/extract_declared_dependencies.py

```diff
@@ -281,6 +322,9 @@ def parse_pyproject_toml(path: Path) -> Iterator[DeclaredDependency]:
 
     yield from parse_pep621_pyproject_contents(parsed_contents, source)
 
+    if "dynamic" in parsed_contents.get("project", {}):
+        yield from parse_dynamic_pyproject_contents(parsed_contents, source)
```

Comment:
Just one quick thing I noticed before I do another proper review:

I believe the dynamic fields are part of PEP621? Or at least more closely related to PEP621's fields than they are to Poetry's fields? In the above lines we _first_ parse the PEP621 fields, and then, if the dynamic fields are present, we _also_ parse them? Shouldn't this be more mutually exclusive? I.e. what happens if a project has both `project.dependencies` and `"dependencies"` in `project.dynamic`? IIUC we should then not look at `project.dependencies` at all?

Reply:
I just re-read @Nour-Mws 's comment again:

> You can have any combination over dependencies or optional-dependencies being dynamic or static.
As soon as one category is declared dynamic, you can no longer declare it statically (but you can still declare the other one statically if not in dynamic).

So if a project has both `project.dependencies` and `"dependencies"` in `project.dynamic`, we should not look at `project.dependencies` at all (which I misunderstood in the last two commits :sweat:  ). To be precise, if a project has `project.dependencies` and `optional-dependencies` in `project.dynamic`, we should look at them both, so `parse_dynamic_pyproject_contents` should be embedded in `parse_pep621_pyproject_contents`?

Reply:
Addressed in https://github.com/tweag/FawltyDeps/pull/351/commits/353ab19dfc899fbd8a2b61e2fbf9100f7133fe3b

Reply:
Yes, I believe the code is correct now, although I'm not 100% sure what happens when `project.dynamic` contains only one of `"dependencies"` or `"other-dependencies"`and not the other. Maybe worth adding two more test cases to make sure we handle all these combinations?

Nitpick alert: Looking at the code in `parse_pep621_pyproject_contents()` and `parse_dynamic_pyproject_contents()` I notice some duplication, specifically in looking up `parsed_contents["project"]["dynamic"]`. Maybe `parse_dynamic_pyproject_contents()` can be refactored into nested functions inside `parse_pep621_pyproject_contents()`, like we already have `parse_main()` and `parse_optional()`? Anyway, If you're so inclined I encourage you to play around with these functions to find the most readable solution.

Reply:
>when project.dynamic contains only one of "dependencies" or "other-dependencies"and not the other. Maybe worth adding two more test cases to make sure we handle all these combinations?

I second that. It would be good to test the following cases:
- `dependencies` in `dynamic` but not `optional-dependencies` (in this case `dependencies` should be read from `["setuptools"]["dynamic"][dependencies"]` and `optional-dependencies` from `["project"]["optional-dependencies"]`)
- the opposite case.